lib/App/Critique/Command/collect.pm

```diff
@@ -179,98 +179,25 @@ sub generate_file_predicate {
-    # ------------------------------#
-    #    0   |   0   |      1       # collect with no violations
-    # ------------------------------#
-
-    # filter only
-    if ( $filter && not($match) && not($no_violation) ) {
-        $predicate = sub {
-            my $root = $_[0];
-            my $path = $_[1];
-            my $rel  = $path->relative( $root );
-            return $rel !~ /$filter/;
-        };
-    }
-    # filter and match
-    elsif ( $filter && $match && not($no_violation) ) {
-        $predicate = sub {
-            my $root = $_[0];
-            my $path = $_[1];
-            my $rel  = $path->relative( $root );
-            return $rel =~ /$match/
-                && $rel !~ /$filter/;
-        };
-    }
-    # filter and match and check violations
-    elsif ( $filter && $match && $no_violation ) {
-        my $c = $session->perl_critic;
-        $predicate = sub {
-            my $root = $_[0];
-            my $path = $_[1];
-            my $rel  = $path->relative( $root );
-            return $rel =~ /$match/
-                && $rel !~ /$filter/
-                && scalar $c->critique( $path->stringify );
-        };
-    }
-    # filter and check violations
-    elsif ( $filter && not($match) && $no_violation ) {
-        my $c = $session->perl_critic;
-        $predicate = sub {
-            my $root = $_[0];
-            my $path = $_[1];
-            my $rel  = $path->relative( $root );
-            return $rel !~ /$filter/
-                && scalar $c->critique( $path->stringify );
-        };
-    }
-    # match only
-    elsif ( not($filter) && $match && not($no_violation) ) {
-        $predicate = sub {
-            my $root = $_[0];
-            my $path = $_[1];
-            my $rel  = $path->relative( $root );
-            return $rel =~ /$match/;
-        };
-    }
-    # match and check violations
-    elsif ( not($filter) && $match && $no_violation ) {
-        my $c = $session->perl_critic;
-        $predicate = sub {
-            my $root = $_[0];
-            my $path = $_[1];
-            my $rel  = $path->relative( $root );
-            return $rel =~ /$match/
-                && scalar $c->critique( $path->stringify );
-        };
-    }
-    # check violations only
-    elsif ( not($filter) && not($match) && $no_violation ) {
-        my $c = $session->perl_critic;
-        $predicate = sub {
-            my $path = $_[1];
-            return scalar $c->critique( $path->stringify );
-        };
-    }
-    # none of the above
-    else {
-        $predicate = sub () { 1 };
-    }
+    my $c = $session->perl_critic;
+
+    # lets build an array of code_ref filters, that will be use to filter
+    # the files, the code refs assume the params will be $path,$rel.
+    my @filters = (sub { return 1 });
+    push @filters, sub { return $_[1] !~ /$filter/} if $filter;
+    push @filters, sub { return $_[1] =~ /$match/ } if $match ;
```

Comment:
I think the order of `filter` and `match` here is reversed, if you look at the old code, `match` always precedes `filter'.